calico_containers/pycalico/datastore.py

```diff
@@ -306,6 +310,31 @@ def remove_host(self, hostname):
             pass
 
     @handle_errors
+    def get_hostnames_from_ips(self, ip_list):
+        """
+        Get the hostnames that are using the given IPs as their calico node IPs.
+        :param ip_list: The list of IPs to get hostnames for.
+        :return: A dictionary of {IP:hostname} the hosts that own the given IPs.
+        """
+        try:
+            hosts = self.etcd_client.read(BGP_HOSTS_PATH, recursive=True)
+            host_ips = hosts.children
+        except EtcdKeyNotFound:
+            # No BGP hosts currently configured in etcd, so no host owns the IP
+            raise KeyError("No BGP host configurations found.")
+
+        ip_host_dict = {}
+
+        # Loop through key-value pairs to find IP addresses
+        for host_ip in host_ips:
+            if host_ip.value in ip_list:
```

Comment:
Surely you need to match the key before matching the value?  A recursive read will return more than just the host BGP IP bindings.